src/pdf.sandbox.external.js

```diff
@@ -0,0 +1,174 @@
+        ) {
+          return;
+        }
+        const id = self.win.setTimeout(() => {
+          self.timeoutIds.delete(callbackId);
+          self.sandbox.callFunction("timeoutCb", {
+            callbackId,
+            interval: false,
+          });
+        }, nMilliseconds);
+        self.timeoutIds.set(callbackId, id);
+      },
+      clearTimeout(id) {
+        self.win.clearTimeout(self.timeoutIds.get(id));
+        self.timeoutIds.delete(id);
+      },
+      setInterval(callbackId, nMilliseconds) {
+        if (
+          typeof callbackId !== "number" ||
+          typeof nMilliseconds !== "number"
+        ) {
+          return;
+        }
+        const id = self.win.setInterval(() => {
+          self.sandbox.callFunction("timeoutCb", {
+            callbackId,
+            interval: true,
+          });
+        }, nMilliseconds);
+        self.timeoutIds.set(callbackId, id);
+      },
+      clearInterval(id) {
+        self.win.clearInterval(self.timeoutIds.get(id));
+        self.timeoutIds.delete(id);
+      },
+      cleanTimeouts() {
+        self.timeoutIds.forEach(([_, id]) => self.win.clearTimeout(id));
+        self.timeoutIds.clear();
+      },
+      alert(cMsg) {
+        if (typeof cMsg !== "string") {
+          return;
+        }
+        if (typeof PDFJSDev !== "undefined" && PDFJSDev.test("TESTING")) {
+          externals.send({ command: "alert", value: cMsg });
+        } else {
+          self.win.alert(cMsg);
+        }
+      },
+      prompt(cQuestion, cDefault) {
+        if (typeof cQuestion !== "string" || typeof cDefault !== "string") {
+          return null;
+        }
+        return self.win.prompt(cQuestion, cDefault);
+      },
+      parseURL(cUrl) {
+        const url = new self.win.URL(cUrl);
+        const props = [
+          "hash",
+          "host",
+          "hostname",
+          "href",
+          "origin",
+          "password",
+          "pathname",
+          "port",
+          "protocol",
+          "search",
+          "searchParams",
+          "username",
+        ];
+
+        return Object.fromEntries(
+          props.map(name => [name, url[name].toString()])
+        );
+      },
+      send(data) {
+        try {
+          if (!data) {
+            return;
+          }
+          const event = new self.win.CustomEvent("updateFromSandbox", {
+            detail: data,
+          });
+          self.win.dispatchEvent(event);
+        } catch (_) {}
+      },
+    };
+
+    return (name, args) => {
+      let result;
+      try {
+        if (args) {
+          args = this.importValueFromSandbox(args);
+          if (!Array.isArray(args)) {
+            result = externals[name](args);
+          } else {
+            result = externals[name](...args);
+          }
+        } else {
+          result = externals[name]();
+        }
+        if (!result) {
+          return null;
+        }
+      } catch (err) {
+        result = { error: err.message };
```

Comment:
The current logic is prone to type confusion, as a return value of `{ error: ... , ... }` is indistinguishable from a caught actual error.

I don't see any matching error-handling (e.g. extracting + re-throwing an error) inside the sandbox either.

In m-c, you can instantiate an error object from the `Cu.andbox` global and throw it (and it will propagate back up into the sandbox).

In quickjs, if you can't do the same, then you'd be forced to do something else, for example serializing and deserializing it.

Reply:
The only function which could throw here is `parseURL` and so it's planned to use the error here and rethrow an exception from its counterpart `crackURL` and then have a correct stack.

Reply:
I don't see how your reply addresses the type confusion that I pointed out.

If you're not doing anything with it right now, I suggest to change it to `throw err;` to get the patch in a reasonable state for merging.